browser/psst/psst_tab_web_contents_observer_unittest.cc

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2025 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/browser/content/psst_tab_web_contents_observer.h"
+
+#include <memory>
+#include <utility>
+
+#include "base/memory/raw_ptr.h"
+#include "base/test/scoped_feature_list.h"
+#include "brave/browser/brave_shields/brave_shields_tab_helper.h"
+#include "brave/components/psst/browser/content/psst_scripts_handler.h"
+#include "brave/components/psst/common/features.h"
+#include "brave/components/psst/common/pref_names.h"
+#include "chrome/browser/prefs/browser_prefs.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/common/chrome_isolated_world_ids.h"
+#include "chrome/test/base/testing_profile.h"
+#include "components/sync_preferences/testing_pref_service_syncable.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_observer.h"
+#include "content/public/test/navigation_simulator.h"
+#include "content/public/test/test_renderer_host.h"
+#include "content/public/test/web_contents_tester.h"
+#include "testing/gmock/include/gmock/gmock.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace psst {
+
+namespace {
+std::unique_ptr<content::BrowserContext> BuildBrowserContext() {
+  TestingProfile::Builder builder;
+  auto prefs = std::make_unique<sync_preferences::TestingPrefServiceSyncable>();
+  ::RegisterUserProfilePrefs(prefs->registry());
+  builder.SetPrefService(std::move(prefs));
+  auto original_profile = builder.Build();
+  return original_profile;
+}
+}  // namespace
+
+class MockPsstScriptsHandler : public PsstScriptsHandler {
+ public:
+  MockPsstScriptsHandler() = default;
+  ~MockPsstScriptsHandler() override = default;
+
+  MOCK_METHOD(void, Start, (), (override));
+};
+
+class PsstTabWebContentsObserverUnitTest
+    : public content::RenderViewHostTestHarness {
+ public:
+  void SetUp() override {
+    feature_list_.InitAndEnableFeature(psst::features::kEnablePsst);
+    content::RenderViewHostTestHarness::SetUp();
+
+    auto script_handler = std::make_unique<MockPsstScriptsHandler>();
+    script_handler_ = script_handler.get();
```

Comment:
this is unsafe, the unique_ptr goes out of scope at the end of `SetUp` and your pointer to `MockPsstScriptsHandler` is not valid after that